[Gabriel Goh] iP

Smanmos opened the pull request:
# BTDuke
> Comes with [BTD6](https://store.steampowered.com/app/960090/Bloons_TD_6/) characters!

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use
All you need to do is,

1. download it.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :flushed: :flushed: :flushed:

And it is **FREE**!

Features:
- [X] Keep track of ToDo, Event, and Deadline
- [X] Good looking GUI
- [ ] Save function (currently bugged)

Commit:
Add Gradle support

Smanmos committed:
Level-1
Add greet, echo, exit functionalities

Smanmos committed:
Level-2
Add add, list functionalities

Smanmos committed:
Level-3

Add mark_done functionality

Smanmos committed:
Add ToDo.java

Smanmos committed:
Add Deadline.java

Smanmos committed:
Create Event.java

Smanmos committed:
Level-4

Add todo, deadline, event functionalities

Smanmos committed:
Add more messages in the program

Smanmos committed:
A-TextUiTesting

Add automated testing

Smanmos committed:
Add error handling

Smanmos committed:
Add delete functionality

Smanmos committed:
Implement file reading

Add file reading features, including support for Tasks to handle additional constructors for done tasks.

Smanmos committed:
Add csv convertion function to Task and subclasses

Smanmos committed:
Correct error in csv function

Add extra type in front so element can be read from file correctly

Smanmos committed:
Add saving functionality to Duke

Allow saving Duke tasks between sessions. This allows the user to keep track of their works even without leaving the program on.

Smanmos committed:
Add timed status to tasks

Currently, tasks with a time (e.g. Event and Deadline) are not differentiated in the code from tasks without a time (e.g. ToDo).

This makes it difficult to add a time formatting function to Duke, as it is difficult to determine which tasks need an extra formatter to print.

By adding new timed property, it becomes easier to print tasks with a time formatter

Smanmos committed:
Add support for recognising different date formats

Currently any string is supported as a time.

This makes it confusing for users, as invalid times are also accepted.

Now strings are validated to make sure the time is formatted correctly, and support for printing the time in different formats is supported.

Smanmos committed:
Resolve merge conflict

Smanmos committed:
Rewrite constructors in Task classes

Constructors has to be rewritten to accomadate changes to base Task class

Smanmos committed:
Rewrite constructors in Task classes

Constructors has to be rewritten to accomadate changes to base Task class

Smanmos committed:
Create new class for UserInterface

Smanmos committed:
Implement UserInterface in Duke partially

Smanmos committed:
Add TaskList class

Isolate function of task list into its own class

Smanmos committed:
Add displayError() to UserInterface

Smanmos committed:
Add Storage loading from file

Smanmos committed:
Fix missing statement in Storage

Smanmos committed:
Move features to other classes

The following functions have been moved:
Obtaining data from storage (moved to Storage)
Getting user input (moved to UserInterface)

Smanmos committed:
Move Duke runner from main into object

Smanmos committed:
Add constructor to Duke

TaoJun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,169 @@
+            }
+            else if (command.startsWith("delete ")){
+                String name = command.substring(7);
+                boolean taskFound = false;
```
Since taskFound is a boolean, maybe taskFound can be renamed to isFound instead?

Siddharth-Sid replied:
> Since taskFound is a boolean, maybe taskFound can be renamed to isFound instead?
I agree according to the naming convention,, maybe taskFound can be renamed to isFound.

TaoJun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,169 @@
+            }
+            else if (command.startsWith("dateformat ")){
+                try {
+                    String newDtformat = command.substring(11);
```
Maybe newDateTimeFormat will be clearer?

TaoJun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,169 @@
+                }
+            }
+            else {
+                List<String> command_list = Arrays.asList("todo", "deadline", "event", "delete");
```
Since this is a collection, maybe commands might be better than command_list?

Smanmos replied:
Thank you for the suggestion!

However, due to the reorganisation of the code, the variable has been removed in future commits.

TaoJun99 reviewed (commented):
Overall, the code quality is good with some potential improvements 👍

jianoway commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDate;
```
Possibly include all your files in a package?

jianoway commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,169 @@
+import java.util.List;
+import java.util.ArrayList;
+
+import java.io.File;
```
Perhaps you could consider revising the order of the imports?

jianoway replied:
Perhaps you might want to reorder them due to a possible lexicographical ordering issue?

jianoway commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,30 @@
+import java.text.DateFormat;
```
Possibly consider using packages to group your files?

jianoway reviewed (commented):
Good code quality overall with some minor improvements such as consistency in naming of your boolean variables as Tao Jun has commented on (: 👍🏻

Smanmos committed:
Add commands to Duke

Currently, commands are all handled within the main Duke runner.

This makes it harder to isolate different commands from Duke.

By adding a Command class, it makes it easier to isolate new features or
discovered bugs to the Commands.

Smanmos committed:
Fix bug where assigned variables were method variables instead of object variables

Smanmos committed:
Add getTasks() method to TaskList

Smanmos committed:
Add saving functionality to Storage

Smanmos committed:
Rewrite Duke.java to shift saving functionality to Storage

Smanmos committed:
Make CommandDone and CommandShowList extend Command

Siddharth-Sid commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,157 @@
+        duke.run();
+    }
+
+    UserInterface ui;
```
Maybe you can mention the access specifier i.e. private

Siddharth-Sid reviewed (commented):
I think there are very few coding style errors are I approve the PR. Request you to take a look at the comments.

Smanmos committed:
Add hasMoreCommands flag to Command.java

Smanmos committed:
Add CommandDelete class

Smanmos committed:
Add CommandExit class

Smanmos committed:
Add Parser helper class

Smanmos committed:
Mark all instance variables of Duke as private

Smanmos committed:
Move command parsing functionality from Duke to Parser

Smanmos committed:
Fix bug in Command

shouldExecuteNext() always returns false.

It now properly returns the hasMoreCommands flag.

Smanmos committed:
Add done command to Parser

Smanmos committed:
Improve parsing of results of command

Allow strings with spaces for todo, event, and deadline commands, and validate input by checking for /at or /by.

Smanmos committed:
Handle multi token strings for done and delete commands

Smanmos committed:
Handle multi token strings for done and delete commands

Smanmos committed:
Move all files to Duke package

Smanmos committed:
Revert changed error messages in Parser from previous commit

Smanmos committed:
Add tests for Deadline

Smanmos committed:
Add tests for ToDo

Smanmos committed:
Add new Find command

Smanmos committed:
Handle find command in Parser

Smanmos committed:
Merge branch 'branch-Level-9'

Smanmos committed:
Fix brace for if block

Smanmos committed:
Fix brace for if block

Smanmos committed:
Wrap long lines

Smanmos committed:
Make a copy of the list of tasks returned

Smanmos committed:
Write JavaDoc for Command

Smanmos committed:
Write JavaDoc for TaskList

Smanmos committed:
Write JavaDoc for Task

Smanmos committed:
Write JavaDoc for CommandAdd

Smanmos committed:
Write JavaDoc for Duke

Smanmos committed:
Write JavaDoc for UserInterface

Smanmos committed:
Write JavaDoc for Parser

Smanmos committed:
Merge branch 'branch-A-CodingStandard'

Smanmos committed:
Merge branch 'branch-A-JavaDoc'

Smanmos committed:
Merge branch 'add-gradle-support' of https://github.com/kawaiigabrielneko/ip into branch-A-Gradle

Smanmos committed:
Merge branch 'branch-A-Gradle'

Smanmos committed:
Add javafx to gradle

Smanmos committed:
Make JavaFX Application

Smanmos committed:
Improve UI

Smanmos committed:
Add button on-press function

Smanmos committed:
Add auto scroll

Smanmos committed:
Add Dialog Box to Duke

Smanmos committed:
Flip dialog box for Duke's response

Smanmos committed:
Add padding for DialogBox

Smanmos committed:
Add spacing between each image and its dialog within each dialog box

Smanmos committed:
Clip each picture so they look like circles

Smanmos committed:
Add background color to each dialog box

Smanmos committed:
Use FXML

Smanmos committed:
Add controller classes

Smanmos committed:
Shift main class to Main

Smanmos committed:
Add getResponse back to Duke

Smanmos committed:
Finish JavaFX tutorial

Smanmos committed:
Convert MainWindow to use fx:root construct

Smanmos committed:
Remove redundant try block

Smanmos committed:
Move some classes into packages

Smanmos committed:
Resolve merge commit

Smanmos committed:
Move DialogBox to ui package

Smanmos committed:
Move MainWindow into ui package

Smanmos committed:
Pass MainWindow into UserInterface

Smanmos committed:
Add images

Smanmos committed:
Update image locations

Smanmos committed:
Update package of DialogBox

Smanmos committed:
Use GUI for input instead of command line

Known issues: Cannot exit; Does not show greeting message

Smanmos committed:
Merge branch 'branch-Level-10'

Smanmos committed:
Add assert to Task

Smanmos committed:
In delete() of TaskList, use find()

Currently, delete() uses code very similar to find() in order to find a targetted Task.

This is an unnecessary duplication of code.

Let's use find() to find the task to delete in delete() to clean up the code, and to adhere to DRY principle.

Smanmos committed:
Implement and use toString() of TaskList

Some classes currently print of tasks by first obtaining the list of individual Tasks from TaskList.

This is bad, because a function that is done on TaskList should be performed by TaskList itself.

Using the SLAP principle, lets move getting the string of tasks into TaskList itself, and printing that to the UI directly instead.

Smanmos committed:
Get Duke to save tasks when it closes

Smanmos committed:
Merge branch 'master' into branch-A-CodeQuality

Smanmos committed:
SLAP out finding tasks with matching keywords from CommandFind into TaskList

Smanmos committed:
Move logic for coverting a TaskList into a csv file from Storage into TaskList

Smanmos committed:
Remove unused functions

Smanmos committed:
Remove unused functions

Smanmos committed:
Extract out main parts of Parser to smaller methods

Smanmos committed:
Change variable name for consistency

Smanmos committed:
SLAP Storage load function

Smanmos committed:
Remove magic constants from Parser and Storage

Smanmos committed:
Remove unused code from UserInterface

Smanmos committed:
Remove magic constants from UserInterface

Smanmos committed:
Clean up imports

Smanmos committed:
Rewrite JavaDoc comments to adhere to coding standard

Smanmos committed:
Merge pull request #2 from kawaiigabrielneko/branch-A-CodeQuality

Improve code quality

Smanmos committed:
Merge branch 'master' of https://github.com/kawaiigabrielneko/ip into branch-A-Assertions

Smanmos committed:
Add assertions to Duke class

Smanmos committed:
Merge pull request #3 from kawaiigabrielneko/branch-A-Assertions

Add assertions

Smanmos committed:
Disallow adding tasks with duplicate names to TaskList.

Currently, duplicate tasks are allowed to be stored.

This can cause issues when referring to Tasks, as they are referred in tasks by name.

Let's reject adding duplicate tasks in TaskList.

Smanmos committed:
Update JavaDoc to reflect change to TaskList

Smanmos committed:
Add try-catch block to Duke.respond

Currently there is no try catch block, because Commands cannot throw exceptions.

This has changed, because CommandAdd will now throw an exception when adding a duplicate task.

Let's surround the body of this method with a try block, and print errors in the catch block.

Smanmos committed:
Resolve merge conflict

Smanmos committed:
Merge pull request #4 from kawaiigabrielneko/branch-C-DetectDuplicates

Reject adding duplicates to TaskList

Smanmos committed:
Correct class and package name

Smanmos committed:
Fix ToDo isDone not saving bug

Smanmos committed:
Reorder methods

Smanmos committed:
Print welcome message when starting app

Smanmos committed:
Stop cutting off long text even if it is too long

Smanmos committed:
Print delete message in one dialog

Smanmos committed:
Print errors differently

Smanmos committed:
Add product representative image

Smanmos committed:
Merge branch 'master' of https://github.com/kawaiigabrielneko/ip

Smanmos committed:
Set a title to the window

Smanmos committed:
Write list command for User Guide

Smanmos committed:
Move Ui.png

Smanmos committed:
Merge branch 'master' of https://github.com/kawaiigabrielneko/ip

Smanmos committed:
Move Ui.png to docs

Smanmos committed:
Add todo command to README.md

Smanmos committed:
Use user-centric phrasing for list command

Smanmos committed:
Add deadline command to README.md

Smanmos committed:
Add event command to README.md

Smanmos committed:
Add done command to README.md

Smanmos committed:
Add delete command to README.md

Smanmos committed:
Add find command to README.md

Smanmos committed:
Add features to README.md

Smanmos committed:
Add second person pronoun to features

Smanmos committed:
Separate format from usage example

Smanmos committed:
Add guide to usage section
